src/Map.ts

```diff
@@ -0,0 +1,825 @@
+ */
+const partitionWithKey = <K, A>(fa: Map<K, A>, p: (k: K, a: A) => boolean): Separated<Map<K, A>, Map<K, A>> => {
+  const left = new Map<K, A>()
+  const right = new Map<K, A>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    if (p(k, a)) {
+      right.set(k, a)
+    } else {
+      left.set(k, a)
+    }
+  }
+  return {
+    left,
+    right
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+const filterMapWithKey = <K, A, B>(fa: Map<K, A>, f: (k: K, a: A) => Option<B>): Map<K, B> => {
+  const m = new Map<K, B>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    const o = f(k, a)
+    if (o.isSome()) {
+      m.set(k, o.value)
+    }
+  }
+  return m
+}
+
+/**
+ * @since 1.14.0
+ */
+const filterWithKey = <K, A>(fa: Map<K, A>, p: (k: K, a: A) => boolean): Map<K, A> => {
+  const m = new Map<K, A>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    if (p(k, a)) {
+      m.set(k, a)
+    }
+  }
+  return m
+}
+
+/**
+ * Create a Map from a foldable collection of key/value pairs, using the
+ * specified function to combine values for duplicate keys.
+ *
+ * @since 1.14.0
+ */
+export function fromFoldable<K, F extends URIS3>(
+  S: Setoid<K>,
+  F: Foldable3<F>
+): <U, L, A>(ta: Type3<F, U, L, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F extends URIS2>(
+  S: Setoid<K>,
+  F: Foldable2<F>
+): <L, A>(ta: Type2<F, L, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F extends URIS>(
+  S: Setoid<K>,
+  F: Foldable1<F>
+): <A>(ta: Type<F, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F>(
+  S: Setoid<K>,
+  // tslint:disable-next-line: deprecation
+  F: Foldable<F>
+): <A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F>(
+  S: Setoid<K>,
+  // tslint:disable-next-line: deprecation
+  F: Foldable<F>
+): <A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A> {
+  return <A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => {
+    return F.reduce<[K, A], Map<K, A>>(ta, new Map<K, A>(), (b, [k, a]) => {
+      const bOpt = lookupWithKey(S)(k, b)
+      b.set(k, bOpt.isSome() ? f(bOpt.value[1], a) : a)
+      return b
+    })
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+const compactable: Compactable2<URI> = {
+  URI,
+  compact,
+  separate
+}
+
+/**
+ * @since 1.14.0
+ */
+const functor: Functor2<URI> = {
+  URI,
+  map: map_
+}
+
+/**
+ * @since 1.14.0
+ */
+export const getFunctorWithIndex = <K>(): FunctorWithIndex2C<URI, K, K> => {
+  return {
+    _L: phantom,
+    ...functor,
+    mapWithIndex: mapWithKey
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const getFoldable = <K>(O: Ord<K>): Foldable2v2C<URI, K> => {
```

Comment:
`getFoldable`, `getFoldableWithIndex`, `getTraversable`, `getTraversableWithIndex` can be merged into `getTraversableWithIndex`

Reply:
I can't do this, because:

- `getTraversable` is also needed for `getWitherable`
- `getFoldable` is needed by both `getTraversable` and `getFoldableWithIndex`

Reply:
Yeah, I misspoke, I mean more "let's keep the public API surface as small as possible", we can export only `getTraversableWithIndex`

Reply:
Ok, I minimized it as much as I could. Things like `getWitherable` and `getFilterableWithIndex` still must be public though.